elastic/apm-server
# Fix race condition in custom libbeat instrumentation

## Motivation/summary

The PR fixes race condition in libbeat instrumentation that surfaces during reload (initial and on subsequent configuration change). The race condition is in setting up custom libbeat instrumentation by using beats `monitoring#DefaultRegistry`.  The race arises from two reload events captured by APM-Server for every reload operation including the initial config load. The two reloads are used by APM-Server to load `inputs` using the beats reload registry and `output` using beats `OutputConfigReloader`.

Specifically, the race condition is in `serverRunner#newFinalBatchProcessor` (called from `serverRunner#run`) due to updates to a global `monitoring#Default` registry and the two reloads explained above. Even though the access to the `monitoring#Default` is protected by mutex in `serverRunner#newFinalBatchProcessor`; it is still possible that `serverRunner#newFinalBatchProcessor` call for a temporary reload happens later for a reload event that happens earlier. This will cause `monitoring#Default` to be in an incorrect state:
1. For initial load, the incorrect state would be no custom instrumentation
2. For subsequent reloads, the incorrect state would be instrumentation with an incorrect model indexer

## Checklist

~~- [ ] Update [package changelog.yml](https://github.com/elastic/apm-server/blob/main/apmpackage/apm/changelog.yml) (only if changes to `apmpackage` have been made)~~
~~- [ ] Documentation has been updated~~

## How to test these changes

1. Run `TestFleetIntegrationMonitoring` multiple times and observe they are not flaky
2. APM-Server metrics for elasticsearch output events should always be present in stack monitoring

## Related issues
Closes #8383
